token/program-2022/src/extension/mod.rs

```diff
@@ -1166,64 +1166,57 @@ impl Extension for AccountPaddingTest {
     const TYPE: ExtensionType = ExtensionType::AccountPaddingTest;
 }
 
-/// Packs a variable-length extension into a new TLV space
-///
-/// This function reallocates the account as needed to accommodate for the
-/// change in space, then initializes the extension.
-pub fn alloc_and_serialize<S: BaseState, V: Extension + VariableLenPack>(
-    account_info: &AccountInfo,
-    extension: &V,
-) -> Result<(), ProgramError> {
-    let previous_account_len = account_info.try_data_len()?;
-    let new_account_len = {
-        let data = account_info.try_borrow_data()?;
-        let state = StateWithExtensions::<S>::unpack(&data)?;
-        state.try_get_new_account_len(extension)?
-    };
-
-    if previous_account_len < new_account_len {
-        // size increased, so realloc the account first
-        account_info.realloc(new_account_len, false)?;
-    }
-
-    let mut buffer = account_info.try_borrow_mut_data()?;
-    // write the account type if needed, so that the next unpack works
-    if previous_account_len <= BASE_ACCOUNT_LENGTH {
-        set_account_type::<S>(*buffer)?;
-    }
-
-    // now alloc in the TLV buffer and write the data
-    let mut state = StateWithExtensionsMut::<S>::unpack(&mut buffer)?;
-    state.init_variable_len_extension(extension, false)
-}
-
-/// Packs a variable-length extension into an existing TLV space
+/// Packs a variable-length extension into a TLV space
 ///
 /// This function reallocates the account as needed to accommodate for the
 /// change in space, then reallocates in the TLV buffer, and finally writes the
 /// bytes.
-pub fn realloc_and_serialize<S: BaseState, V: Extension + VariableLenPack>(
+///
+/// NOTE: This function does not protect against double-alloc, and will realloc
+/// an existing extension. The calling code must check if the extension already
+/// exists, and return an error if appropriate.
```

Comment:
I'm fine with this approach as-is, but if you think we need more guardrails, we could pass an `overwrite: bool` like we do with `StateWithExtensionsMut::init_extension()` to make things more obvious and controllable. Wdyt?

Reply:
Do you mean something like:

```
if extension_already_exists && !overwrite {
    return Err(...);
}
```

That seems reasonable to me!

Reply:
Yes, exactly!